securedrop/tests/migrations/migration_48a75abc0121.py

```diff
@@ -0,0 +1,383 @@
+        VALUES (:reply_id, :journalist_id)
+        """
+        try:
+            db.engine.execute(text(sql), **params)
+        except IntegrityError:
+            pass
+
+    @staticmethod
+    def mark_file_as_seen(file_id, journalist_id):
+        params = {
+            "file_id": file_id,
+            "journalist_id": journalist_id,
+        }
+        sql = """
+        INSERT INTO seen_files (file_id, journalist_id)
+        VALUES (:file_id, :journalist_id)
+        """
+        try:
+            db.engine.execute(text(sql), **params)
+        except IntegrityError:
+            pass
+
+    @staticmethod
+    def mark_message_as_seen(message_id, journalist_id):
+        params = {
+            "message_id": message_id,
+            "journalist_id": journalist_id,
+        }
+        sql = """
+        INSERT INTO seen_messages (message_id, journalist_id)
+        VALUES (:message_id, :journalist_id)
+        """
+        try:
+            db.engine.execute(text(sql), **params)
+        except IntegrityError:
+            pass
+
+
+class UpgradeTester(Helper):
+    """
+    This migration verifies that the seen_files, seen_messages, and seen_replies association tables
+    now exist, and that the data migration completes successfully.
+    """
+
+    JOURNO_NUM = 20
+    SOURCE_NUM = 20
+
+    def __init__(self, config):
+        self.config = config
+        self.app = create_app(config)
+
+    def load_data(self):
+        with self.app.app_context():
+            for _ in range(self.SOURCE_NUM):
+                self.add_source()
+
+            for _ in range(self.JOURNO_NUM):
+                self.add_journalist()
+
+            for i in range(self.SOURCE_NUM):
+                # add 1-3 messages from each source, some messages are set to downloaded
+                for _ in range(random.randint(1, 3)):
+                    self.add_message(i)
+                # add 0-2 files from each source, some files are set to downloaded
+                for _ in range(random.randint(0, 2)):
+                    self.add_file(i)
+
+            # add 30 replies from randomly-selected journalists to randomly-selected sources
+            for i in range(30):
+                selected_journo = random.randint(0, self.JOURNO_NUM)
+                selected_source = random.randint(0, self.SOURCE_NUM)
+                self.add_reply(selected_journo, selected_source)
+
+    def check_upgrade(self):
+        with self.app.app_context():
+            sql = "SELECT * FROM submissions"
+            submissions = db.engine.execute(text(sql)).fetchall()
+
+            sql = "SELECT * FROM replies"
+            replies = db.engine.execute(text(sql)).fetchall()
+
+            # Now seen tables exist, so you should be able to mark some files, messages, and replies
+            # as seen
+            for submission in submissions:
+                if submission.filename.endswith('-doc.gz.gpg') and random.choice([0, 1]):
```

Comment:
These `random.choice` gates could mean you don't end up marking anything as seen, if you get really unlucky. 

It looks like the mark-as-seen helper functions don't do anything that you're not testing immediately after, when checking for unique constraint violations, so do we need these `for` loops and the helper methods at all?

Reply:
yup, in the comment is says: `Now seen tables exist, so you should be able to mark some files, messages, and replies as seen`. i just wanted to mark some files, messages, and replies as seen (not all in th db)

> It looks like the mark-as-seen helper functions don't do anything that you're not testing immediately after, when checking for unique constraint violations, so do we need these for loops and the helper methods at all?

the helper functions are to make it easier to create journalists, submissions, seen records, etc. and to read what's going on. the for loop is to loop through all submissions so that we can make sure we are able to add seen records once there are seen tables

Reply:
I understand what you're trying to do, but it might not be happening _at all_ in the for loops, because of the `random.choice` calls. You're not verifying the results of the for loops, and you're testing the same thing, the basic ability to insert to the seen tables, immediately below during the unique constraint checks. 

I would either remove the for loops altogether, or be more deliberate about the subset of records you mark seen, and then test that the inserted seen table records are as expected.

The helper functions are fine; you could also use them in the unique constraint testing.

Reply:
You know, I just realized that because you are testing the insertion during the unique constraint check, it doesn't matter at all if the `random.choice` calls skip all that work. It's actually better, because it could save time whenever this migration is run, which is a lot in dev environments.